sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,902 @@
+
+        Parameters
+        ----------
+        X : array-like, shape (n_query, n_features)
+            Test samples.
+
+        Returns
+        -------
+        y_pred : array, shape (n_query,)
+            A predicted class label for each test sample.
+
+        Raises
+        ------
+        NotFittedError
+            If :meth:`fit` has not been called before.
+        """
+
+        # Check if fit has been called
+        check_is_fitted(self, ['L_'])
+        y_pred = super(LargeMarginNearestNeighbor, self).predict(
+            self.transform(X))
+
+        return y_pred
+
+    def predict_proba(self, X):
+        """Return probability estimates for the test data X.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_query, n_features)
+            Test samples.
+
+        Returns
+        -------
+        p : array of shape = [n_samples, n_classes], or a list of n_outputs
+            of such arrays if n_outputs > 1.
+            The class probabilities of the input samples. Classes are ordered
+            by lexicographic order.
+
+        Raises
+        ------
+        NotFittedError
+            If :meth:`fit` has not been called before.
+        """
+
+        # Check if fit has been called
+        check_is_fitted(self, ['L_'])
+        probabilities = super(LargeMarginNearestNeighbor, self).predict_proba(
+            self.transform(X))
+
+        return probabilities
+
+    def _validate_params(self, X, y):
+        """Validate input parameters as soon as :meth:`fit` is called.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features_in)
+            The training samples.
+
+        y : array-like, shape (n_samples,)
+            The corresponding training labels.
+
+        Returns
+        -------
+        X : array, shape (n_samples, n_features_in)
+            The training samples.
+
+        y_inversed : array, shape (n_samples,)
+            The corresponding training labels.
+
+        Raises
+        -------
+        TypeError
+            If a parameter's type does not match the desired type.
+
+        ValueError
+            If a parameter's value violates its legal value range or if the
+            combination of two or more given parameters is incompatible.
+        """
+
+        # Check training data
+        X, y = check_X_y(X, y, ensure_min_samples=2)
+        check_classification_targets(y)
+
+        # Store the appearing classes and the class index for each sample
+        classes, y_inversed = np.unique(y, return_inverse=True)
+
+        # Check number of classes > 1
+        n_classes = len(classes)
+        if n_classes < 2:
+            raise ValueError("LargeMarginNearestNeighbor requires 2 or more "
+                             "distinct classes, got {}.".format(n_classes))
+
+        # Check every class has at least 2 samples
+        min_class_size = bincount(y_inversed).min()
```

Comment:
I tried doing that by setting the label of singletons to something negative like -2, but then many parts of the rest of the algorithm also have to change, as it is based on indexing of the original inputs. Right now I make a copy of the non-singleton inputs to train LMNN, which I know is ugly.. I'll try to find an alternative.